Add ability to hide projects

arsduo opened the pull request:
Our chapter, like many others, gets a lot of applications that aren't valid -- people asking for money to fix their homes, start a business, or just submitting the same application twice. This seems to have been a problem for as long as AF has been going (issue #37 was opened back in 2012).

After talking it over with the Chicago chapter, this pull request implements an "Hide" button. If a trustee or dean sees a grant they don't think is valid, they can save the rest of their chapter time by clicking on "Hide" and entering a reason they don't think the grant is valid.

To keep everything transparent, hidden applications don't disappear completely -- they show up in abbreviated form in the list with who hid them, when, and why. If another trustee disagrees, they can hit the "unhide" button. Additionally, if you click through to the individual project page, you can see all the details for the project.

Screenshots:

_Prompt shows up for each entry:_
<img width="1258" alt="bildschirmfoto 2015-10-10 um 16 32 15" src="https://cloud.githubusercontent.com/assets/48325/10413404/4a28c716-6f6d-11e5-9f80-8207fa7420fa.png">

_Entering a reason to hide:_
<img width="1258" alt="bildschirmfoto 2015-10-10 um 16 32 29" src="https://cloud.githubusercontent.com/assets/48325/10413406/61311a8a-6f6d-11e5-9c37-51455b604193.png">

_After hiding, you're brought back to the same spot and can see the hidden project:_
<img width="1258" alt="bildschirmfoto 2015-10-10 um 16 33 28" src="https://cloud.githubusercontent.com/assets/48325/10413411/7a85e7a4-6f6d-11e5-8a63-6157499675c8.png">

_In the individual view, you get the full details:_
<img width="1258" alt="bildschirmfoto 2015-10-10 um 16 33 33" src="https://cloud.githubusercontent.com/assets/48325/10413416/8e485484-6f6d-11e5-8ddb-80e22e3774d7.png">

_If you unhide, you see the whole project again:_
<img width="1258" alt="bildschirmfoto 2015-10-10 um 16 33 37" src="https://cloud.githubusercontent.com/assets/48325/10413419/95d2e2be-6f6d-11e5-966b-86ea14974c16.png">

arsduo commented:
A few implementation notes:

* In our chapter, we'd want everyone to be able to hide this. If other chapters felt strongly we could potentially add an optional gating so only the deans could hide.
* In #37, there was discussion of how this would work if "another chapter might want to fund something that has been passed over by the "intended" chapter". Now that we have several years of experience, is this still a concern? 
  * If some chapters do look at projects other chapters received, does that happen enough -- and do we expect projects to be hidden inappropriately enough -- to merit figuring out a solution for that case?
* The reason isn't required, but I can add some validation at the controller level if desired.

Any feedback would be very welcome. Thanks!

jcn commented on app/views/projects/_project.html.erb:
```diff
@@ -1,84 +1,110 @@
+    <% if project.archived? %>
+      <section class="meta-data">
+        <div>
+          Archived by <%= project.archived_by_user.name %> on <%= project.archived_at %> because <%= project.archived_reason %>.
```
Please add this to `en.yml`

jcn commented:
Overall, this looks decent, and I naturally have some initial thoughts (I will actually check the implementation in a bit):

* Make sure that all of the text is localizable. I commented on the one spot I saw, but check for others (I didn't see any, but it doesn't mean I didn't miss them as well)
* I think it's fine for this to be available to anyone, instead of deans-only
* To that end though when you click the "hide" link, let's add some language about a) this will hide the project for everyone and b) that this is reversible
* What was the thinking behind putting the link at the top of the project instead of the bottom? My first inclination would be the bottom, after you'd read it (I'd actually love to move this eventually to an "activity" menu, maybe like #43, but that's not necessary for this PR)
* Why the discrepancy between "hide" in the interface and "archive" in the db? We should probably be consistent (if the names are different it probably means that during development, the intent changed)

arsduo commented:
I'd swapped out archive for hide everywhere but it looks I forgot to push
that up — sorry! I'll do that and make the other suggested changes tomorrow.
On Oct 10, 2015 6:06 PM, "Jesse Chan-Norris" <notifications@github.com>
wrote:

> Overall, this looks decent, and I naturally have some initial thoughts (I
> will actually check the implementation in a bit):
>
>    - Make sure that all of the text is localizable. I commented on the
>    one spot I saw, but check for others (I didn't see any, but it doesn't mean
>    I didn't miss them as well)
>    - I think it's fine for this to be available to anyone, instead of
>    deans-only
>    - To that end though when you click the "hide" link, let's add some
>    language about a) this will hide the project for everyone and b) that this
>    is reversible
>    - What was the thinking behind putting the link at the top of the
>    project instead of the bottom? My first inclination would be the bottom,
>    after you'd read it (I'd actually love to move this eventually to an
>    "activity" menu, maybe like #43
>    <https://github.com/awesomefoundation/awesomebits/issues/43>, but
>    that's not necessary for this PR)
>    - Why the discrepancy between "hide" in the interface and "archive" in
>    the db? We should probably be consistent (if the names are different it
>    probably means that during development, the intent changed)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/awesomefoundation/awesomebits/pull/202#issuecomment-147133595>
> .
>

arsduo commented:
I've pushed up some updates:

* The feature is called hiding throughout.
* The form is at the bottom of the entry.
* All text goes through i18n (and the hidden details are a single i18n string with variables, allowing for better translations).
* The "Hide" button now also hides the "Not an appropriate project?" prompt when showing the form.
* You have to supply a reason, otherwise you get an appropriate flash message. (You're not redirected back to the specific project in order to show it, unfortunately.)

Let me know what you think!

_Hiding is now at the bottom_
<img width="1389" alt="bildschirmfoto 2015-10-10 um 23 25 53" src="https://cloud.githubusercontent.com/assets/48325/10414710/61ed601a-6fa8-11e5-9f9b-66c37044614c.png">

_The prompt to hide is hidden when viewing the form, which contains an explanation_
<img width="1389" alt="bildschirmfoto 2015-10-10 um 23 25 59" src="https://cloud.githubusercontent.com/assets/48325/10414711/6be182f4-6fa8-11e5-996c-fd3c99c793de.png">

_The explanation is internationalized (and still works)_
<img width="1389" alt="bildschirmfoto 2015-10-10 um 23 31 35" src="https://cloud.githubusercontent.com/assets/48325/10414715/a8d3ae1c-6fa8-11e5-9539-b39c5645f9e2.png">

_When you fail to provide a reason, you get a flash warning and the project isn't hidden_
<img width="1389" alt="bildschirmfoto 2015-10-10 um 23 39 37" src="https://cloud.githubusercontent.com/assets/48325/10414725/b8583ce0-6fa8-11e5-90eb-90e7cb7499f6.png">
<img width="1389" alt="bildschirmfoto 2015-10-10 um 23 39 42" src="https://cloud.githubusercontent.com/assets/48325/10414726/ba0e6992-6fa8-11e5-81b9-0f456baf1078.png">

emdaniels commented:
This is so excellent! I know the flow through grant reading for many trustees will really benefit from this new feature. I'm looking forward to using it! One thing I noticed- the Unhide button is the same pink as when you click Awesome on a project (except without the outline). It makes a weird association in my mind that the hidden project is sort of awesome. Could we change the Unhide button color to gray? It wouldn't stand out as much then. What do others think?

arsduo commented:
Great idea! I've just pushed an update that visually differentiates the two buttons by making hide/unhide gray:

<img width="734" alt="bildschirmfoto 2015-10-11 um 10 17 11" src="https://cloud.githubusercontent.com/assets/48325/10417443/65275fac-7001-11e5-8af7-0d9ee56446f3.png">
<img width="758" alt="bildschirmfoto 2015-10-11 um 10 16 29" src="https://cloud.githubusercontent.com/assets/48325/10417442/6526e518-7001-11e5-8433-a3c5953f6edf.png">

It doesn't have any hover behavior, which is a little inconsistent with the other button. I can try later to make it invert like the other buttons (becoming gray on black?) if that's worth doing.

jcn commented on app/controllers/projects_controller.rb:
```diff
@@ -85,6 +85,23 @@ def update
+      @project.hide!(params[:hidden_reason], current_user)
+      redirect_to chapter_projects_path(@project.chapter_id, anchor: "project#{@project.id}")
+    else
+      flash[:notice] = "You must supply a reason a to hide a project!"
```
Here's another spot where we need to use a translated string.

jcn commented:
@arsduo That border around the hide button is inconsistent with our other buttons. We have the `filter` button that has the hover state you were looking for, and we have this `@button` scss mixin that you should use. I actually think the layout still looks a bit off, but I'm willing to let that go at least for this pass just to ship it. 

Also, in actually reviewing this code, two things jump out at me:

1. It seems like the `full_view` boolean when rendering the view is backwards - shouldn't `full_view` be the default and if we want to hide the project, we should set that to false? Or, `full_view` should be called `hidden` in which case it can default to false.
2. Given we rip so much of the display out when a project is hidden, I'm tempted to say there should be two partials, a `_project_details` and a `_project_hidden` which I think will clean it up and make it clear from the main render what is actually happening. We can still render `_project` for the header and some of the buttons, but then the remainder can just be a render of one of these views. I don't want to get too granular, but `if/else` blocks in partials can get really dicey and people can end up sticking the logic in the wrong place - I find that breaking things into logical partials makes it more readable. 

Thoughts?

avicaplan commented:
Love it! Thanks so much for building this. I do wonder though if the logical place to put the "Hide" and "Unhide" buttons might actually be to the left of the "Awesome" button. Since that's also where the "Winner for..." button appears for deans, it seems to me to be the place for toggling things about the project.

arsduo commented:
Sorry about the delay in responding — it's been a busy week and I will get back to this PR next week. 




Jesse — good points on the code and style, wil make those changes.




Avi — I don't have strong feelings either way visually, but it's marginally easier to put it below. That said, if the consensus is up top that's easy to do – but if so, where should we show the explanatory text?

On Wed, Oct 14, 2015 at 10:34 PM, avicaplan <notifications@github.com>
wrote:

> Love it! Thanks so much for building this. I do wonder though if the logical place to put the "Hide" and "Unhide" buttons might actually be to the left of the "Awesome" button. Since that's also where the "Winner for..." button appears for deans, it seems to me to be the place for toggling things about the project.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/awesomefoundation/awesomebits/pull/202#issuecomment-148272224

avicaplan commented:
I actually don't think the explanation of what hiding does is strictly necessary, and it's just one more thing to localize anyways.

For the prompt to input a reason, I don't know what's technically possible but something that pops up and floats near the button would seem sensible to me.

For hidden applications, my inclination would be to keep the "Hidden by..." at the bottom, as you've done it -- though I think there is little value in including the date and time and we could just say "Hidden by... because [reason]." Or, this may work more cleanly in other languages when localized if it simply reads "Hidden by... . [Reason]."

Just my two cents! I'm happy with this feature in any form, really.

jcn commented:
@avicaplan some notes:

> Since that's also where the "Winner for..." button appears for deans, it seems to me to be the place for toggling things about the project.

I don't actually like that big pile of buttons anyway (which I what I was saying above) so I don't want to clutter that area with one MORE button at this point. At least for now I'd like to introduce "hiding" more subtly like Alex did.

> I actually don't think the explanation of what hiding does is strictly necessary, and it's just one more thing to localize anyways.

We can always display the text in English until it gets localized. I think that the idea that it's hidden for everyone is something important to convey.

> For the prompt to input a reason, I don't know what's technically possible but something that pops up and floats near the button would seem sensible to me.

I'm not sure what you're hoping to get out of this. Why do you want this behavior vs what's there now?

> For hidden applications, my inclination would be to keep the "Hidden by..." at the bottom, as you've done it -- though I think there is little value in including the date and time and we could just say "Hidden by... because [reason]." Or, this may work more cleanly in other languages when localized if it simply reads "Hidden by... . [Reason]."

We can fix the structure a bit, but at least for me, as Dean, that is information that I'd like to know - was this hidden by someone in the chapter? Recently? Who did it? Can I talk to them about it? Especially since over time, things change, maybe when I revisit I'd like to know. If you want to tweak the localized language a bit, that's a different story. `This project is hidden: [reason] (Jesse Chan-Norris on [date] )` or something. But I really don't want to get too caught up in the details since we'll probably have to make changes anyway.

avicaplan commented:
Ok, makes sense! And having the text input float would only be necessary if the button was alongside the "Awesome" button, so no need to worry about that suggestion now.

Maybe we can leave the date if you think it's important, but not bother with the time.

arsduo commented:
Hey all! I've updated the display (the button now matches the filter button) and have pulled out the project detail display into two different partials.

Let me know if you have any additional feedback! I'll have time to work on it this week. It'd be neat to get it live for the November grants if possible.

![bildschirmfoto 2015-10-25 um 18 26 50](https://cloud.githubusercontent.com/assets/48325/10718671/54ed2298-7b46-11e5-9f6e-19080ec6eaf8.png)
![bildschirmfoto 2015-10-25 um 18 26 42](https://cloud.githubusercontent.com/assets/48325/10718672/5653fa30-7b46-11e5-8a19-2a78770163c2.png)

avicaplan commented:
Would be happy to see this rolled out as is, but two ideas from the peanut gallery for making it take even less screen real estate: could it be included in the same horizontal grey bar as "Edit" appears in (or does that bar only appear for deans)? And why not just go with an "Unhide" link in the same style as the "Hide" one?

Also, I think "at" should be "on." :)

arsduo commented:
The spec failure is a translation issue -- I just have to update the files having moved some links around. I'll do that later tonight.

arsduo commented:
I've fixed the spec failure (updating en.yml to account for the new views). Everything seems to be working in my tests.

![bildschirmfoto 2015-10-25 um 21 16 57](https://cloud.githubusercontent.com/assets/48325/10719824/fbbe2682-7b5d-11e5-974f-5bcb7e9d8881.png)
![bildschirmfoto 2015-10-25 um 21 16 36](https://cloud.githubusercontent.com/assets/48325/10719825/fbc0377e-7b5d-11e5-81e5-ddb00c34e3d9.png)
![bildschirmfoto 2015-10-25 um 18 26 50](https://cloud.githubusercontent.com/assets/48325/10719826/fbc1a686-7b5d-11e5-88ab-d008627756ef.png)

@avicaplan -- thanks for catching at -> on! That should've changed when I stopped showing the time.

Putting it in the gray bar would require more work, since we'd have to figure out what to do with the explanation. I'm not a strong designer, so I'm not the best person to tackle that; if someone tells me how to do it I can adjust that, but I'd suggest we defer that for later (if someone wants to do a design pass separately).

arsduo commented:
@jcn I have time to work on this today if there are any additional changes to be merge. If possible it'd be great to get it in before it's time review grants for November!

jcn commented:
Thanks for the ping, Alex. I'm planning an awesomebits working session for myself for later today and hope to be able to merge this in at that time. 


> On Nov 1, 2015, at 11:47 AM, Alex Koppel <notifications@github.com> wrote:
> 
> @jcn I have time to work on this today if there are any additional changes to be merge. If possible it'd be great to get it in before it's time review grants for November!
> 
> —
> Reply to this email directly or view it on GitHub.
>

jcn commented on Rakefile:
```diff
@@ -15,3 +15,9 @@ if defined?(RSpec)
+task :migrate => :environment do
+  Rake::Task["db:migrate"].execute
+  Rake::Task["db:test:prepare"].execute
+end
```
Can you explain this block of code?

arsduo replied:
Oops, I should've removed that. I added it to help me with development; it combines migrating the development database and updating the test DB so they're always in sync. I'll remove it; if it seems useful I can submit a separate pull request to add it.

On Sun, Nov 1, 2015 at 12:32 PM, Jesse Chan-Norris
<notifications@github.com> wrote:

>> @@ -15,3 +15,9 @@ if defined?(RSpec)
>>      end
>>    end
>>  end
>> +
>> +desc "db:migrate + db:test:prepare"
>> +task :migrate => :environment do
>> +  Rake::Task["db:migrate"].execute
>> +  Rake::Task["db:test:prepare"].execute
>> +end
> Can you explain this block of code?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/awesomefoundation/awesomebits/pull/202/files#r43585631

jcn commented on app/models/project.rb:
```diff
@@ -192,6 +194,26 @@ def extra_answer(num)
+    update_attributes(
+      hidden_reason: reason,
+      hidden_by_user_id: user.id,
+      hidden_at: Time.now
```
We should use Time.zone.now so the time stored will be correct regardless of what the system time zone is set as. 

http://winstonyw.com/2014/03/03/time_now_vs_time_zone_now/

arsduo replied:
Pushed, thanks for catching this!

jcn commented on config/locales/en.yml:
```diff
@@ -242,4 +253,4 @@ en:
               invalid: must be lowercase and can not contain spaces
   meta:
-    description: Forwarding the interest of Awesome in the universe, $1,000 at a time.
+    description: Forwarding the interest of Awesome in the universe, $1,000 at a time.
```
Missing a newline at the end of this file.

arsduo replied:
Added. Just out of curiosity, is there a solid reason to prefer having newlines over not, or is it just convention? (Convention is totally fine; we don't do this at eSpark and I'm curious if there's a reason to.)

jcn commented on config/locales/en.yml:
```diff
@@ -102,19 +102,30 @@ en:
+      edit: Edit Project →
+    project_hidden:
+      confirm-delete: Really delete this?
+      view-public-page: View the public page →
```
This duplication of translation strings seems weird. Why was this broken out into these two sections? If it's because of lazy interpolation, I would instead make the calls as `t(".view-public-page", :scope => "projects.project")` and then use the same translation in both places. Since these are actually the same string being displayed in the same place (not just the same translation of 2 different strings) I think this is more correct.

arsduo replied:
Perfect. I couldn't actually find documentation on how to change the scope (clearly didn't look hard enough), hence the documentation. I'll update that in an hour or so so there's only one copy of each.

On Sun, Nov 1, 2015 at 12:49 PM, Jesse Chan-Norris
<notifications@github.com> wrote:

>>        about-me: "A little about me:"
>>        about-project: "Here's my idea:"
>>        use-for-money: "How I will use the money:"
>>        not-answered: Not Answered
>> -      images: "Images:"
>> -      image: "Image #%{index}"
>> +      confirm-delete: Really delete this?
>> +      view-public-page: View the public page →
>> +      edit: Edit Project →
>> +    project_hidden:
>> +      confirm-delete: Really delete this?
>> +      view-public-page: View the public page →
> This duplication of translation strings seems weird. Why was this broken out into these two sections? If it's because of lazy interpolation, I would instead make the calls as `t(".view-public-page", :scope => "projects.project")` and then use the same translation in both places. Since these are actually the same string being displayed in the same place (not just the same translation of 2 different strings) I think this is more correct.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/awesomefoundation/awesomebits/pull/202/files#r43585821

jcn commented:
My comment about the translation might seem nitpicky, but our translation process is a little complicated right now, and I try to do everything to prevent our translators more pain than is necessary. In this case, moving around a lot of translation strings and duplicating them will lead to more confusion. I'll add this to the docs / wiki, but as a rule if we're adding new strings in a given section we should add them to the end of that section to make it easier to find and translate. And if you're going to have to restructure the i18n yaml hierarchy, then you're going to have to do it in all of the language files (not just `en.yml`). In this particular case, I think it makes sense to use the `t` function's scoping option and to leave the translated strings in the top-level of the `projects.project` scope.

jcn commented:
Another comment (and I hope these aren't too annoying - I truly appreciate the functionality here and think it's a great feature): how big a pain would it be to do the hiding/unhiding in an Ajax call. From a user interaction perspective, reloading the page on every hide, especially in the use case you describe where there are a lot of projects to hide in a given month, is going to be tedious for the hider. In addition, with pagination, if you're on page 2 and you click hide you're going to be bounced back to page 1, which is going to be a jarring experience. The code for marking a winner and shortlisting a project is a pretty good mechanism I think, and could apply pretty well:

https://github.com/awesomefoundation/awesomebits/blob/master/app/assets/javascripts/projects.js.coffee

The only issue is that the way those work is they apply a class to the surrounding div which changes the behavior of that div. I think this might work here, but it would break the other thing I'd asked you to do earlier with breaking the display into 2 partials. 

This is, of course, a failure on my part to properly review the code in the first place, and I apologize for that. I can take a stab at it myself and make the appropriate changes if it's too annoying to have to revert these changes. I'm just thinking that from a user-perspective, this makes for a much better experience.

Also happy to discuss on slack if that'd be easier.

jcn commented:
> The only issue is that the way those work is they apply a class to the surrounding div which changes the behavior of that div.

In playing with this locally, it seems like a couple of more changes would be appropriate:

If an application is hidden, the user gets the opportunity to review it before unhiding it. Clicking the "View application" button just unhides the body. 
![screen shot 2015-11-01 at 2 30 48 pm](https://cloud.githubusercontent.com/assets/2909/10870913/dbb35e96-80a5-11e5-8b7a-a0b0512cd794.png)

The actual unhide button is displayed once the application is shown:
![screen shot 2015-11-01 at 2 31 45 pm](https://cloud.githubusercontent.com/assets/2909/10870914/dbb3f69e-80a5-11e5-845b-1d3dbd7db240.png)

My feeling about the "awesome," "winner" and "edit" links is that those will all be moving around eventually anyway (I kind of hate where they all live) so if it makes things easier to just have them all displayed for both the hidden and unhidden versions, the we just go with that. The hiding and unhiding can just hide the application itself (which is the bulk of the ugliness anyway) and we can focus on moving around the other interface elements later. 

How does that sound?

arsduo commented:
I like your suggestion of only showing the unhide button in the full view; I'll make that change. I'll also see if I can get the unhide to go to the right page, that shouldn't be hard.



I initially looked into building this with ajax, but not being familiar enough with the site I  decided to go in a direction I knew I could finish in the time I have. It'd certainly be possible to make it dynamic, but that's honestly not something I can take on in the next few weeks. 



I appreciate the feedback — it's not annoying at all. It's worth the extra effort to figure out what the right solution is, given how much more work it takes to change code once it's merged in. It seems like there are two good options at this point — wait and get this working with ajax or finish this with the current design and refactor it along with the rest of the buttons (since it seems like that's on the horizon). What do you think?

On Sun, Nov 1, 2015 at 1:40 PM, Jesse Chan-Norris
<notifications@github.com> wrote:

>> The only issue is that the way those work is they apply a class to the surrounding div which changes the behavior of that div.
> In playing with this locally, it seems like a couple of more changes would be appropriate:
> If an application is hidden, the user gets the opportunity to review it before unhiding it. Clicking the "View application" button just unhides the body. 
> ![screen shot 2015-11-01 at 2 30 48 pm](https://cloud.githubusercontent.com/assets/2909/10870913/dbb35e96-80a5-11e5-8b7a-a0b0512cd794.png)
> The actual unhide button is displayed once the application is shown:
> ![screen shot 2015-11-01 at 2 31 45 pm](https://cloud.githubusercontent.com/assets/2909/10870914/dbb3f69e-80a5-11e5-845b-1d3dbd7db240.png)
> My feeling about the "awesome," "winner" and "edit" links is that those will all be moving around eventually anyway (I kind of hate where they all live) so if it makes things easier to just have them all displayed for both the hidden and unhidden versions, the we just go with that. The hiding and unhiding can just hide the application itself (which is the bulk of the ugliness anyway) and we can focus on moving around the other interface elements later. 
> How does that sound?
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/awesomefoundation/awesomebits/pull/202#issuecomment-152856060

arsduo commented:
As a side note, could you invite me to the slack channel? I've also never managed to get onto the trustee mailing list; do you know who manages that?

On Sun, Nov 1, 2015 at 1:09 PM, Jesse Chan-Norris
<notifications@github.com> wrote:

> Another comment (and I hope these aren't too annoying - I truly appreciate the functionality here and think it's a great feature): how big a pain would it be to do the hiding/unhiding in an Ajax call. From a user interaction perspective, reloading the page on every hide, especially in the use case you describe where there are a lot of projects to hide in a given month, is going to be tedious for the hider. In addition, with pagination, if you're on page 2 and you click hide you're going to be bounced back to page 1, which is going to be a jarring experience. The code for marking a winner and shortlisting a project is a pretty good mechanism I think, and could apply pretty well:
> https://github.com/awesomefoundation/awesomebits/blob/master/app/assets/javascripts/projects.js.coffee
> The only issue is that the way those work is they apply a class to the surrounding div which changes the behavior of that div. I think this might work here, but it would break the other thing I'd asked you to do earlier with breaking the display into 2 partials. 
> This is, of course, a failure on my part to properly review the code in the first place, and I apologize for that. I can take a stab at it myself and make the appropriate changes if it's too annoying to have to revert these changes. I'm just thinking that from a user-perspective, this makes for a much better experience.
> Also happy to discuss on slack if that'd be easier.
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/awesomefoundation/awesomebits/pull/202#issuecomment-152852540

arsduo commented:
I've addressed all feedback:

* hiding applications now returns you to exact application on the appropriate page (unless you're doing it from the detail view, in which case it returns you to page 1 -- I think this is a minor enough case that we can wait to fix it when we remove the page reload).
* the unhide button only shows up in the detail page (and clicking it returns you to that page)
* removed duplication in the translations

This will definitely be improved a lot by using ajax. The Chicago chapter is going to try to do quarterly hackathons to work on the site, and could potentially tackle this (and any UI redesign) with you next time.

Hiding something on page 2:
![bildschirmfoto 2015-11-01 um 15 03 05](https://cloud.githubusercontent.com/assets/48325/10871390/36b7acd4-80ab-11e5-9b89-2b44d2df8d54.png)

Goes back to the application on p2 with no unhide button:
![bildschirmfoto 2015-11-01 um 15 03 58](https://cloud.githubusercontent.com/assets/48325/10871391/39bec336-80ab-11e5-825e-72c09e4229ce.png)

That's visible in the detail page:
![bildschirmfoto 2015-11-01 um 15 15 31](https://cloud.githubusercontent.com/assets/48325/10871395/6b6d98d0-80ab-11e5-8793-1e2c1a08b543.png)

And unhiding takes you back to the detail page:
![bildschirmfoto 2015-11-01 um 15 15 35](https://cloud.githubusercontent.com/assets/48325/10871396/6c7540a2-80ab-11e5-9eee-3c31768a65fd.png)

jcn commented:
Totally fine with leaving the ajax until later (after we merge in). I've invited you to slack, and we can discuss mailing list stuff there. I will review the code later tonight and assuming everything looks good enough will merge in.

jcn commented:
@arsduo Sorry, can't merge this in yet. There are a couple of problems related to translations:

1. A hidden project has this weird HTML issue due to the translation being missing when rendering a translation on the index page:

![screen shot 2015-11-02 at 7 49 40 am](https://cloud.githubusercontent.com/assets/2909/10882216/ecb4b690-8136-11e5-9b24-63bf7c0a3654.png)

2. The project page just completely craps out in non-english languages because they don't have a `default` date set (which is being called by the localization of the `.to_date` call)

3. The project pages will be incomplete as well because you moved around all of the translation strings into the new `project_details` key in the `en.yml` file but not any of the others.

To fix this:

1. Add a `default` date translation key to all of the locale files (it can be the same for all of them but it has to exist)

2. Move your translations keys for `en.yml` back to their original namespace (and their original order) and simply add your 8 new keys for `unhide`, `consider-hiding` etc to that same namespace. You'll have to use scoping in the `t()` calls, but it'll make it much easier to do the translations. As of now we will have to redo the namespaces and ordering of keys in all of the other translation files in order to get these pages to display properly now. And considering we're talking about re-doing the buttons, etc later anyway, I don't think that the translation strings warrant moving around just for the sake of the partials (which is what's happening when you put the translations in the `project_details` namespace).

arsduo commented:
Aha, I was viewing that as a regular trustee and missed the issue with the
Dean link. (That's odd, though, because there's a test for that.)

I'll dig into these issues tonight and push up a fix.
On Nov 2, 2015 07:06, "Jesse Chan-Norris" <notifications@github.com> wrote:

> @arsduo <https://github.com/arsduo> Sorry, can't merge this in yet. There
> are a couple of problems related to translations:
>
>    1. A hidden project has this weird HTML issue due to the translation
>    being missing when rendering a translation on the index page:
>
> [image: screen shot 2015-11-02 at 7 49 40 am]
> <https://cloud.githubusercontent.com/assets/2909/10882216/ecb4b690-8136-11e5-9b24-63bf7c0a3654.png>
>
>    1.
>
>    The project page just completely craps out in non-english languages
>    because they don't have a default date set (which is being called by
>    the localization of the .to_date call)
>    2.
>
>    The project pages will be incomplete as well because you moved around
>    all of the translation strings into the new project_details key in the
>    en.yml file but not any of the others.
>
> To fix this:
>
>    1.
>
>    Add a default date translation key to all of the locale files (it can
>    be the same for all of them but it has to exist)
>    2.
>
>    Move your translations keys for en.yml back to their original
>    namespace (and their original order) and simply add your 8 new keys for
>    unhide, consider-hiding etc to that same namespace. You'll have to use
>    scoping in the t() calls, but it'll make it much easier to do the
>    translations. As of now we will have to redo the namespaces and ordering of
>    keys in all of the other translation files in order to get these pages to
>    display properly now. And considering we're talking about re-doing the
>    buttons, etc later anyway, I don't think that the translation strings
>    warrant moving around just for the sake of the partials (which is what's
>    happening when you put the translations in the project_details
>    namespace).
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/awesomefoundation/awesomebits/pull/202#issuecomment-153011508>
> .
>

arsduo commented:
I've updated the PR, and I think resolved most of the issues. For some reason, though, Rails i18n isn't falling back to the English keys when translations are missing -- any help in figuring that out would be awesome.

_explanation inline for the unhide to save space_
![bildschirmfoto 2015-11-10 um 19 41 06](https://cloud.githubusercontent.com/assets/48325/11081700/a49a500e-87e4-11e5-8b05-7d11164ff63e.png)

_no translation errors_
![bildschirmfoto 2015-11-10 um 19 36 50](https://cloud.githubusercontent.com/assets/48325/11081702/aa640ac0-87e4-11e5-8477-f1bb75058e80.png)

_problems with translations_
![bildschirmfoto 2015-11-10 um 19 53 55](https://cloud.githubusercontent.com/assets/48325/11081712/cf69b19e-87e4-11e5-81f8-3f77fda637d3.png)

jcn commented:
@arsduo It looks like the fallback will work properly in production ([see code](https://github.com/awesomefoundation/awesomebits/blob/master/config/environments/production.rb#L64-L66)), it's just disabled in development so it's obvious to you where you're missing translations. We'll get the translators to add language after people have had a chance to try it out and we tweak any language, etc that might need to happen.

arsduo commented:
@jcn just updated the pull request with the changes you made earlier. Sorry it took a while -- looking forward to wrapping this up at last!

jcn commented:
There are a couple of more things that I'd like to tweak with this, but for now I'll merge it in so we can see how it performs in production.
